x/vote/keeper/migration.go

```diff
@@ -0,0 +1,143 @@
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/axelarnetwork/axelar-core/utils"
+	evmtypes "github.com/axelarnetwork/axelar-core/x/evm/types"
+	"github.com/axelarnetwork/axelar-core/x/vote/exported"
+	"github.com/axelarnetwork/axelar-core/x/vote/exported2"
+	"github.com/axelarnetwork/axelar-core/x/vote/types"
+	"github.com/axelarnetwork/axelar-core/x/vote/types2"
+)
+
+// GetMigrationHandler returns the migration handler for the vote module
+func GetMigrationHandler(k Keeper) func(ctx sdk.Context) error {
+	return func(ctx sdk.Context) error {
+		return migrateVotes(ctx, k)
+	}
+}
+
+func migrateVotes(ctx sdk.Context, k Keeper) error {
+	metadatas := k.getPollMetadatasOld(ctx)
+	for _, metadata := range metadatas {
+		newMetadata := exported.PollMetadata{
+			Key:              metadata.Key,
+			ExpiresAt:        metadata.ExpiresAt,
+			Result:           MigrateVoteData(k.cdc, metadata.Result, k.Logger(ctx)),
+			VotingThreshold:  metadata.VotingThreshold,
+			State:            metadata.State,
+			MinVoterCount:    metadata.MinVoterCount,
+			Voters:           metadata.Voters,
+			TotalVotingPower: metadata.TotalVotingPower,
+			RewardPoolName:   metadata.RewardPoolName,
+		}
+
+		pollStore := k.newPollStore(ctx, newMetadata.Key)
+		pollStore.SetMetadata(newMetadata)
+		votes := pollStore.GetVotesOld()
+		for _, vote := range votes {
+			newVote := types.TalliedVote{
+				Tally:  vote.Tally,
+				Voters: vote.Voters,
+				Data:   MigrateVoteData(k.cdc, vote.Data, k.Logger(ctx)),
+			}
+			pollStore.Set(votesPrefix.AppendStr(pollStore.key.String()).AppendStr(newVote.Hash()), &newVote)
+		}
+	}
+
+	return assertMigrationSuccessful(ctx, k)
+}
+
+func assertMigrationSuccessful(ctx sdk.Context, k Keeper) error {
+	metadatas := k.getPollMetadatas(ctx)
+	for _, metadata := range metadatas {
+		if metadata.Result.GetCachedValue() == nil {
+			return fmt.Errorf("failed to verify poll result for %s", hex.EncodeToString(k.cdc.MustMarshalLengthPrefixed(&metadata)))
+		}
+
+		pollStore := k.newPollStore(ctx, metadata.Key)
+		votes := pollStore.GetVotes()
+		for _, vote := range votes {
+			if metadata.Result.GetCachedValue() == nil {
+				return fmt.Errorf("failed to verify tallied vote data for %s", hex.EncodeToString(k.cdc.MustMarshalLengthPrefixed(&vote)))
+			}
+		}
+	}
+	return nil
+}
+
+// MigrateVoteData migrates vote results from an Any slice to a single Any value
+func MigrateVoteData(cdc codec.BinaryCodec, data *codectypes.Any, logger log.Logger) *codectypes.Any {
+	switch d := data.GetCachedValue().(type) {
+	case *exported2.Vote:
+		events, err := unpackEvents(cdc, d.Results)
+		if err != nil {
+			logger.Error("failed tp unpack vote results for %s", hex.EncodeToString(cdc.MustMarshalLengthPrefixed(d)))
+			return nil
+		}
+
+		result, err := evmtypes.PackEvents(events[0].Chain, events)
```

Comment:
`events` may be empty here, correct me if I'm wrong @haiyizxx ? In which case I don't think `.Chain` is available anymore.